### PR TITLE
Add Controller gRPC

### PR DESCRIPTION
**Pull request by Langenbahn**
## Change Summary
Added RPC declarations to `controller.proto`
Added implementation of `ProbeCommunicatorServer` to controller in `rpc.go`
Added concurrent timeout checker to handle the case in which a probe becomes decoupled from the controller
General refactoring to improve readability and reduce the complexity of control flow

Updated test cases and adding RPC logic to the probe will both come in separate PRs.

### CODE REVIEW

File: Controller/src/controller/regionalVM.go [hunk 1]
```diff
@@ -20,48 +20,80 @@ import (
...
-func (vm *regionalVM) startProbes(acc *AccountInfo) {
-	s := &ProbeConfigs{Probe: vm.probes}
-	p := proto.MarshalTextString(s)
-	a := proto.MarshalTextString(acc)
+func (vm *regionalVM) stopVM() {
+	/*err := maker.Command("gcloud", "compute", "instances", "delete", vm.name, "--zone", vm.zone, "--quiet").Run()
+	if err != nil {
+		log.Printf("stopVM: unable to stop VM %s in zone %s: %v", vm.name, vm.zone, err)
+	}*/
+	log.Print("Stopping VM")
```
Comment (RobertMParker) [inline Controller/src/controller/regionalVM.go:65]: I would change this message to be "Stopping VM - Not implemented"

File: Controller/src/controller/rpc.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	err := makeCert()
+	if err != nil {
+		return err
+	}
+	tls, err := credentials.NewServerTLSFromFile("cert.pem", "key.pem")
+	if err != nil {
+		return err
+	}
+	srv := grpc.NewServer(grpc.Creds(tls))
+	lis, err := net.Listen("tcp", "crdhost.c.gifted-cooler-279818.internal:50001")
```
Comment (RobertMParker) [inline Controller/src/controller/rpc.go:31]: Don't hardcode this address, it should be a flag

File: Controller/src/controller/rpc.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return err
+	}
+	controller.RegisterProbeCommunicatorServer(srv, new(CommunicatorServer))
+}
+
+func makeCert() error {
+	err := maker.Command("openssl", "req", "-x509", "-newkey", "rsa:4096", "-keyout", "key.pem", "-out", "cert.pem", "-days", "365", "-nodes", "-subj", "'/CN=crdhost.c.gifted-cooler-279818'").Run()
```
Comment (RobertMParker) [inline Controller/src/controller/rpc.go:39]: same here

File: Controller/src/controller/rpc.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return err
+	}
+	// Read public certificate so that it can be included in regional vm metadata
+	cert, err := ioutil.ReadFile("cert.pem")
```
Comment (RobertMParker) [inline Controller/src/controller/rpc.go:44]: make "cert.pem" a constant

File: Controller/src/controller/controller.go [hunk 5]
```diff
@@ -23,65 +23,73 @@ package controller
...
 
-var maker utils.CommandMaker
+var (
+	maker          utils.CommandMaker
+	vms            map[string]*regionalVM
+	stoppedVMs     int32
+	stoppedVMsLock sync.Mutex
+	running        bool
+	config         *ControllerConfig
+)
```
Comment (brotherlogic2) [inline Controller/src/controller/controller.go:39]: Not quite sure why these aren't in the struct?

File: Controller/src/controller/regionalVM.go [hunk 6]
```diff
@@ -20,48 +20,80 @@ import (
 	"log"
 )
 
+type vmState int
+
+const (
+	inactive state = iota
```
Comment (brotherlogic2) [inline Controller/src/controller/regionalVM.go]: You've got vmState and state here?

File: Controller/src/controller/regionalVM.go [hunk 7]
```diff
@@ -20,48 +20,80 @@ import (
...
-	err := maker.Command("gcloud", "compute", "instances", "delete", vm.name).Run()
-	if err != nil {
-		log.Printf("stopVM: unable to stop VM %s in zone %s", vm.name, vm.zone)
-		return err
+func (vm *regionalVM) updatePingTime() {
+	vm.lastPing = time.Now()
+}
+
+func (vm *regionalVM) setState(s vmState) {
+	vm.stateLock.Lock()
```
Comment (brotherlogic2) [inline Controller/src/controller/regionalVM.go:88]: you can stick a defer vm.stateLock.Unlock() here and remove below

File: Controller/src/controller/rpc.go (hunk 3 above)
Comment (brotherlogic2) [reply Controller/src/controller/rpc.go]: I would also indent this so the flag and value are on the same line in code:

maker.Command("openssl", "req",
 "-x509",
" -newkey", "rsa:4096"...)

etc. Makes it a bit easier to read.

File: Controller/src/controller/rpc.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	vms[in.Source].updatePingTime()
+	in.Source = "Controller"
+	in.Stop = running
+	return in, nil
+}
+
+func checkVMs(max time.Duration, wg *sync.WaitGroup) {
+	for stoppedVMs < len(vms) {
+		for _, v := range vms {
+			vm.stateLock.Lock()
```
Comment (brotherlogic2) [inline Controller/src/controller/rpc.go:104]: might be better to put this in a function so you can defer the Unlock - I think that gives you and us a bit more confidence that we're not forgetting to unlock a lock.

File: Controller/src/controller/rpc.go (hunk 4 above)
Comment (Langenbahn) [reply Controller/src/controller/rpc.go]: Addressed in 85540dd

File: Controller/src/controller/rpc.go (hunk 3 above)
Comment (Langenbahn) [reply Controller/src/controller/rpc.go]: Addressed in 85540dd

File: Controller/src/controller/rpc.go (hunk 2 above)
Comment (Langenbahn) [reply Controller/src/controller/rpc.go]: Addressed in 85540dd

File: Controller/src/controller/regionalVM.go (hunk 1 above)
Comment (Langenbahn) [reply Controller/src/controller/regionalVM.go]: Addressed in 85540dd

File: Controller/src/controller/regionalVM.go (hunk 6 above)
Comment (Langenbahn) [reply Controller/src/controller/regionalVM.go]: Addressed in 85540dd

File: Controller/src/controller/controller.go (hunk 5 above)
Comment (Langenbahn) [reply Controller/src/controller/controller.go:39]: The variables in the struct don't have global scope, so they would have to be explicitly passed to functions when they are needed, or through other functions if they're needed in nested function calls.

File: Controller/src/controller/regionalVM.go (hunk 7 above)
Comment (Langenbahn) [reply Controller/src/controller/regionalVM.go:88]: Addressed in d2c167e

File: Controller/src/controller/rpc.go (hunk 8 above)
Comment (Langenbahn) [reply Controller/src/controller/rpc.go]: addressed in d2c167e

**Pull request merged**